proxy/util/gcloudutil.go

```diff
@@ -0,0 +1,119 @@
+// Copyright 2018 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package util
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"os/exec"
+	"runtime"
+	"time"
+
+	"github.com/GoogleCloudPlatform/cloudsql-proxy/logging"
+	"golang.org/x/oauth2"
+)
+
+// GcloudConfigData represents the data returned by `gcloud config config-helper`.
+type GcloudConfigData struct {
+	Configuration struct {
+		Properties struct {
+			Core struct {
+				Project string
+				Account string
+			}
+		}
+	}
+	Credential struct {
+		AccessToken string    `json:"access_token"`
+		TokenExpiry time.Time `json:"token_expiry"`
+	}
+}
+
+func (cfg *GcloudConfigData) oauthToken() *oauth2.Token {
+	return &oauth2.Token{
+		AccessToken: cfg.Credential.AccessToken,
+		Expiry:      cfg.Credential.TokenExpiry,
+	}
+}
+
+type GcloudStatusCode int
+
+const (
+	GcloudOk GcloudStatusCode = iota
+	GcloudNotFound
+	// generic execution failure error not specified above.
+	GcloudExecErr
+)
+
+type GcloudError struct {
+	GcloudError error
+	Status      GcloudStatusCode
+}
+
+func (e *GcloudError) Error() string {
+	return e.GcloudError.Error()
+}
+
+// GcloudConfig returns a GcloudConfigData object or an error of type *GcloudError.
+func GcloudConfig() (*GcloudConfigData, error) {
+	gcloudCmd := "gcloud"
+	if runtime.GOOS == "windows" {
+		gcloudCmd = gcloudCmd + ".cmd"
+	}
+
+	if _, err := exec.LookPath(gcloudCmd); err != nil {
+		return nil, &GcloudError{err, GcloudNotFound}
+	}
+
+	buf := new(bytes.Buffer)
+	cmd := exec.Command(gcloudCmd, "--format", "json", "config", "config-helper")
```

Comment:
Found some issue of this.

This `gcloud` command doesn't seem to respect `GOOGLE_APPLICATION_CREDENTIALS`. I have `Google Cloud SDK 221.0.0`.

So assuming the following setup: user starts the proxy in GCE/GKE with `GOOGLE_APPLICATION_CREDENTIALS` set and that points to a service account in a different project; user does not use `-credential_file` because according to https://cloud.google.com/sql/docs/mysql/sql-proxy#environment, it `is identical to supplying a credential file on the command line`. However in this case GCE default service account will be used to start the proxy and it may not have proper permission of the cloud sql instance.

Either this is an unintended result, or documentation needs to be adapted.

Reply:
This was unintended, but I'll update the documentation instead of reverting it.

Reply:
It seems the order described in doc is still not reflecting code:

```
Credentials supplied in the proxy invocation command
Credentials supplied by the local environment
Credentials associated with the Compute Engine instance
Credentials from an authenticated Cloud SDK client
```

Reply:
#229